### ISSUE TITLE
Many-to-many associations throw "TypeError: Cannot convert null to object" error on create and populate

### ISSUE THREAD

**Issue by gnestor** [balderdashy/sails-postgresql#130]
### Problem
When I try to associate one collection with another collection using a many-to-many association, I get the following error: 

```
TypeError: Cannot convert null to object
    at hasOwnProperty (native)
    at utils.object.hasOwnProperty (/Users/grant/Sites/sails/test/node_modules/sails-postgresql/node_modules/waterline-sequel/sequel/lib/utils.js:28:14)
    at /Users/grant/Sites/sails/test/node_modules/sails-postgresql/node_modules/waterline-sequel/sequel/where.js:261:11
    at Array.forEach (native)
    at WhereBuilder.complex (/Users/grant/Sites/sails/test/node_modules/sails-postgresql/node_modules/waterline-sequel/sequel/where.js:178:36)
    at complexWhere (/Users/grant/Sites/sails/test/node_modules/sails-postgresql/node_modules/waterline-sequel/sequel/index.js:245:16)
    at find (/Users/grant/Sites/sails/test/node_modules/sails-postgresql/node_modules/waterline-sequel/sequel/index.js:85:23)
    at Cursor.populateBuffers [as $populateBuffers] (/Users/grant/Sites/sails/test/node_modules/sails-postgresql/lib/adapter.js:547:31)
    at Cursor.run (/Users/grant/Sites/sails/test/node_modules/sails-postgresql/node_modules/waterline-cursor/cursor/cursor.js:45:8)
    at runJoins (/Users/grant/Sites/sails/test/node_modules/sails-postgresql/node_modules/waterline-cursor/index.js:51:10)
``` 

### Test project
I have created a sample Sails app to demonstrate the issue and allow others to test/debug: https://github.com/gnestor/sails-assoications-issue

### To reproduce:
Try to create a new record with an association (I've tried using Sails console, Blueprints API, and inside the controllers): 

`Event.create({source: 'http://domain.com/event/1', posted_by: {source: 'http://domain.com/feed/1'}}).exec(console.log)`

`Event.create({source: 'http://domain.com/event/3', venue: {source: 'http://domain.com/feed/2'}}).exec(console.log)`

`Event.create({source: 'http://domain.com/event/2', reposted_by: [{source: 'http://domain.com/feed/2'}, {source: 'http://domain.com/feed/4'}]}).exec(console.log)`

### Result

You'll notice that the first two work (one-to-many associations) but the third doesn't (many-to-many association). I have tested this a variety of ways (using Sails console, Blueprints API, and inside controllers). I have also tested with sails-mongo and I get the same results with a slightly different error (specific to sails-mongo):

```
TypeError: Cannot read property 'where' of undefined
    at _afterFetchingJunctorRecords (/Users/grant/Sites/sails/test/node_modules/sails-mongo/node_modules/waterline-cursor/cursor/populateBuffers.js:131:35)
    at /Users/grant/Sites/sails/test/node_modules/sails-mongo/lib/collection.js:103:5
    at /Users/grant/Sites/sails/test/node_modules/sails-mongo/node_modules/mongodb/lib/mongodb/cursor.js:172:16
    at commandHandler (/Users/grant/Sites/sails/test/node_modules/sails-mongo/node_modules/mongodb/lib/mongodb/cursor.js:720:16)
    at /Users/grant/Sites/sails/test/node_modules/sails-mongo/node_modules/mongodb/lib/mongodb/db.js:1874:9
    at Server.Base._callHandler (/Users/grant/Sites/sails/test/node_modules/sails-mongo/node_modules/mongodb/lib/mongodb/connection/base.js:453:41)
    at /Users/grant/Sites/sails/test/node_modules/sails-mongo/node_modules/mongodb/lib/mongodb/connection/server.js:481:18
    at MongoReply.parseBody (/Users/grant/Sites/sails/test/node_modules/sails-mongo/node_modules/mongodb/lib/mongodb/responses/mongo_reply.js:68:5)
    at null.<anonymous> (/Users/grant/Sites/sails/test/node_modules/sails-mongo/node_modules/mongodb/lib/mongodb/connection/server.js:439:20)
    at emit (events.js:95:17)
```

It's worth noting that it does work with sails-disk but not with sails-postgresql, sails-mongo, or sails-mysql, so I'm probably doing something wrong.
 
I have been struggling with this issue for a couple weeks now. I opened a previous issue but I've narrowed the issue down since then: https://github.com/balderdashy/sails-mongo/issues/209

Here's another similar issue: https://github.com/balderdashy/sails-mongo/issues/189

**Comment by tjwebb** [comment]
I can't make sense of your domain objects. follower, post, venue, and all these other things are all of type "Feed"? Your ORM should model real life. It doesn't make logical sense to say that a "venue"  and a "follower" are the same ORM type. I think I can't directly help you with your bug report because I am just as confused as Waterline.

**Comment by gnestor** [comment]
The idea is that events belong to feeds and the user discovers events via feeds. A feed can post an event (Feed.posts), repost an event (Feed.reposts), and--if it's a venue--host an event (Feed.venue_of). A feed can also follow (Feed.following) and be followed by (Feed.followers) other feeds. 

Thanks for your response, Travis. I apparently don't understand something important here. I would REALLY appreciate some clarification or direction.

**Comment by gnestor** [comment]
I'm encountering these errors in `sails console`. I tried using Blueprints API and it worked. I then tried doing `Feed.find().populateAll().exec(...)` inside my controllers and running using `sails lift` and it also worked! So why would `sails console` throw this error and not `sails lift`.

**Comment by tjwebb** [comment]
Is it possible you have a different version of sails installed globally as the version in your local project's node_modules?

I'd love to help, my bottom line is just that I really don't know why you are having this problem. Part of this is that I can't understand your ORM design.

> and--if it's a venue--host an event (Feed.venue_of).
Then where's your `Venue` model? This is what I'm saying: everything seems to be of type "Feed" somehow. I don't understand how it makes sense that a Venue **is a** Feed.

**Comment by tjwebb** [comment]
@gnestor how are things going?

**Comment by gnestor** [comment]
Hey @tjwebb thanks for checking back in. 

My models: It's difficult to explain my reasoning behind my models, but essentially we have feeds and events, and feeds are related to events via their `posts`, `reposts`, and `venue_of` properties. I considered creating a venue model, but venues have the same properties as feeds, only their relationship to events is different (e.g. `venue_of`). 

Problem: It's solved now. For some reason, Sails console will throw these errors above when trying to associate a record involving a many-to-many association or populate a record involving a many-to-many association. When I tested the same queries via Blueprints API or inside my controllers, I stopped getting these errors. 

Closing this issue now... Thanks again for your help, @tjwebb.
